p2p/starknet/iterator.go

```diff
@@ -62,15 +87,36 @@ func (it *iterator) Next() bool {
 	return it.Valid()
 }
 
-func (it *iterator) BlockNumber() uint64 {
-	return it.blockNumber
-}
-
 func (it *iterator) Block() (*core.Block, error) {
-	block, err := it.bcReader.BlockByNumber(it.blockNumber)
+	var (
+		block *core.Block
+		err   error
+	)
+
+	if it.blockHash != nil {
+		block, err = it.bcReader.BlockByHash(it.blockHash)
+		if err == nil {
+			// after first successful fetch by hash
+			// we switch to iteration by block number
+			it.blockHash = nil
+			it.blockNumber = block.Number
+		}
+	} else {
+		block, err = it.bcReader.BlockByNumber(it.blockNumber)
+	}
```

Comment:
IMO, we can avoid this additional complexity. We can keep `iterator` operate on block numbers only and make `newIteratorByHash` look up the  block number of a given block hash before creating the iterator. Similar to how `ByHash` functions work in `blockchain` for example;

```golang
func blockHeaderByHash(txn db.Transaction, hash *felt.Felt) (*core.Header, error) {
	var header *core.Header
	return header, txn.Get(db.BlockHeaderNumbersByHash.Key(hash.Marshal()), func(val []byte) error {
		var err error
		header, err = blockHeaderByNumber(txn, binary.BigEndian.Uint64(val))
		return err
	})
}
```

Reply:
I agree